panther_driver/src/driver_node.py

```diff
@@ -285,51 +325,38 @@ def _roboteq_state_is_correct(self, flags: list) -> bool:
-        msg.undervoltage = bool(flag_val & ref_flag_val << 2)
-        msg.short_circuit = bool(flag_val & ref_flag_val << 3)
-        msg.emergency_stop = bool(flag_val & ref_flag_val << 4)
-        msg.motor_or_sensor_setup_fault = bool(flag_val & ref_flag_val << 5)
-        msg.mosfet_failure = bool(flag_val & ref_flag_val << 6)
-        msg.default_config_loaded_at_startup = bool(flag_val & ref_flag_val << 7)
-
-        return msg
+    def _decode_flag(
+        self, msg_obj: Union[FaultFlag, RuntimeError, ScriptFlag], msg_fields_list: list, flag_val: int
+    ) -> Union[FaultFlag, RuntimeError, ScriptFlag]:
+        
+        if hasattr(msg_obj, 'can_net_err') and self._panther_can.can_connection_error():
+            msg_obj.can_net_err = True
+            rospy.logerr_throttle(
+                5.0, 
+                f'[{rospy.get_name()}] CAN interface connection error (SdoCommunicationError)'
+            )
 
-    @staticmethod
-    def _decode_script_flag(flag_val: int) -> int:
-        ref_flag_val = 0b00000001
+        faults = [
+            (lambda field: (setattr(msg_obj, field, True), field))(field_name)[1]
+            for i, field_name in enumerate(msg_fields_list)
+            if bool(flag_val & 0b00000001 << i)
+        ]
 
-        msg = ScriptFlag()
-        msg.loop_error = bool(flag_val & ref_flag_val << 0)
-        msg.encoder_disconected = bool(flag_val & ref_flag_val << 1)
-        msg.amp_limiter = bool(flag_val & ref_flag_val << 2)
+        if faults and faults != ['safety_stop_active']:
+            self._log_faults(faults)
 
-        return msg
+        return msg_obj
 
-    @staticmethod
-    def _decode_runtime_flag(flag_val: int) -> RuntimeError:
-        # For more info see 272-roboteq-controllers-user-manual-v21 p. 247
-        #   https://www.roboteq.com/docman-list/motor-controllers-documents-and-files/
-        ref_flag_val = 0b00000001
-
-        msg = RuntimeError()
-        msg.amps_limit_active = bool(flag_val & ref_flag_val << 0)
-        msg.motor_stall = bool(flag_val & ref_flag_val << 1)
-        msg.loop_error = bool(flag_val & ref_flag_val << 2)
-        msg.safety_stop_active = bool(flag_val & ref_flag_val << 3)
-        msg.forward_limit_triggered = bool(flag_val & ref_flag_val << 4)
-        msg.reverse_limit_triggered = bool(flag_val & ref_flag_val << 5)
-        msg.amps_trigger_activated = bool(flag_val & ref_flag_val << 6)
-
-        return msg
+    def _log_faults(self, faults: list):
+        faults_str = ', '.join(faults)
+        warn_str = f'[{rospy.get_name()}] Motor controller has detected a fault or runtime error: {faults_str}'
+
+        if set(faults).issubset(self._faults_list):
+            rospy.logerr_throttle(2.0, warn_str)
+        elif set(faults).issubset(self._runtime_errors_list):
+            rospy.logwarn_throttle(2.0, warn_str)
+        elif set(faults).issubset(self._script_flags_list):
+            rospy.logerr_throttle(2.0, warn_str)
```

Comment:
Shouldn't it have sort of like flags memory? If an error occurs once for about 0.1s and second before that there as another error already logged, this will skip logging it. It could, for example, log every time there appears a new flag, but it gets throttled if there was no change in a while.